lib/Sema/CSApply.cpp

```diff
@@ -2400,10 +2400,37 @@ namespace {
                                          cs.getConstraintLocator(base));
         }
 
+        Type toType = simplifyType(expr->getType());
+        
+        // Don't allow lvalues when indexing into a tuple expr.
+        // If we get any lvalues, add load exprs to convert the tuple to be all rvalues.
+        if (auto *tupleExpr = dyn_cast<TupleExpr>(base)) {
+          unsigned count = tupleExpr->getNumElements();
+          unsigned lvalues = 0;
+          auto &tc = cs.getTypeChecker();
+          SmallVector<TupleTypeElt, 4> tupleElts;
+          for (unsigned i = 0; i < count; i++) {
+            Expr *elementExpr = tupleExpr->getElement(i);
+            Type elementType = elementExpr->getType();
+            if (elementType->isLValueType()) {
+              lvalues++;
+              elementExpr->propagateLValueAccessKind(AccessKind::Read, true);
+              elementExpr = new (tc.Context) LoadExpr(elementExpr, elementType->getRValueType());
```

Comment:
I don't know enough about SIL to say absolutely that it won't, but in every other location where a LoadExpr is created, propagateLValueAccessKind() gets called just before on the contained expression (like here), and that would assert if called on an rvalue or on a tuple with a mix of lvalues and rvalues. So certainly no other part of Sema exposes SIL to the possibility.